crates/worker/src/partition_processor_manager/mod.rs

```diff
@@ -702,6 +742,73 @@ impl PartitionProcessorManager {
         }
     }
 
+    fn spawn_create_snapshot_task(
+        &mut self,
+        partition_id: PartitionId,
+        result_sender: oneshot::Sender<anyhow::Result<SnapshotId>>,
+        archived_lsn_sender: watch::Sender<Option<Lsn>>,
+    ) {
+        debug!("In spawn create snapshot task...");
+
+        if self
+            .pending_snapshot_export_tasks
+            .get(&partition_id)
+            .is_some_and(|task| !task.is_finished())
+        {
+            warn!(%partition_id, "Snapshot creation already in progress, rejecting request");
+            result_sender
+                .send(Err(anyhow!("Snapshot creation already in progress")))
+                .ok();
+        } else {
+            let config = self.updateable_config.live_load();
+
+            let snapshot_base_path = config.worker.snapshots.snapshots_dir(partition_id);
+
+            let create_snapshot_task = SnapshotPartitionTask {
+                cluster_name: config.common.cluster_name().into(),
+                node_name: config.common.node_name().into(),
+                partition_id,
+                partition_store_manager: self.partition_store_manager.clone(),
+                snapshot_base_path,
+                result_sender,
+                archived_lsn_sender,
+            };
+            let snapshot_span = tracing::info_span!("create-snapshot");
+
+            let spawn_task_result = restate_core::task_center().spawn_unmanaged(
+                TaskKind::PartitionSnapshotProducer,
+                "create-snapshot",
+                Some(partition_id),
+                async move {
+                    create_snapshot_task
+                        .create_snapshot()
+                        .await
+                        .inspect_err(|err| {
+                            warn!("Unhandled error in create_snapshot task: {}", err)
+                        })
+                }
+                .instrument(snapshot_span),
+            );
+
+            match spawn_task_result {
+                Ok(task) => {
+                    _ = self
+                        .pending_snapshot_export_tasks
+                        .insert(partition_id, task);
+                    debug!("In-progress create snapshto task...");
+                }
+                Err(err) => {
+                    // todo(pavel): how do we solve the ownership of sender moving to the (now failed) task?
+                    // sender.send(Err(anyhow!("Shutting down"))).ok();
```

Comment:
Halp! How do I make this better? The `sender` gets moved to the async block above, but if we're shutting down, spawning fails and I lose the channel to notify the sender.

Reply:
One way how this could be solved is the following: Instead of passing the `sender` and other channel senders to the `SnapshotPartitionTask`, we could let the task only return the snapshot id and the min applied lsn. Then we store the `sender` in `pending_snapshot_export_tasks` and take the `TaskHandle` and poll it until it completes (e.g. adding it to a `FutureUnordered` and polling it as part of the main loop of the PPM). Once it completes we get the result and can send it back to the caller by removing the sender from our map. This has also the benefit that any error that makes the snapshot task fail will be properly reported.

Reply:
Nice! I'll try this approach.